crates/bevy_ecs/src/query/filter.rs

```diff
@@ -512,6 +512,183 @@ macro_rules! impl_tuple_query_filter {
+/// #
+/// # #[derive(Component)]
+/// # struct Position {};
+/// #
+/// fn print_still_entity_system(query: Query<Entity, Not<Changed<Position>>>) {
+///     for entity in &query {
+///         println!("Entity {:?} did not move", entity);
+///     }
+/// }
+/// # bevy_ecs::system::assert_is_system(print_still_entity_system);
+/// ```
+///
+/// Filtering for components that were mutated:
+/// ```
+/// # use bevy_ecs::component::Component;
+/// # use bevy_ecs::entity::Entity;
+/// # use bevy_ecs::query::Added;
+/// # use bevy_ecs::query::Changed;
+/// # use bevy_ecs::query::Not;
+/// # use bevy_ecs::system::IntoSystem;
+/// # use bevy_ecs::system::Query;
+/// #
+/// # #[derive(Component)]
+/// # struct MyComponent;
+/// #
+/// fn print_mutated_component_system(
+///     query: Query<Entity, (Changed<MyComponent>, Not<Added<MyComponent>>)>
+/// ) {
+///     for entity in &query {
+///         println!("Entity {:?} had MyComponent changed, but was not just added", entity);
+///     }
+/// }
+/// # bevy_ecs::system::assert_is_system(print_mutated_component_system);
+/// ```
+pub struct Not<T>(PhantomData<T>);
+
+/// SAFETY:
+///
+/// For purely archetypal filters, this inverts the access in `update_component_access` as well as in
+/// `matches_component_set`, ensuring that they match.
+///
+/// For purely non-archetypal filters, this maintains the same read accesses as the original filter in
+/// both instances.
+
+unsafe impl<T> WorldQuery for Not<T>
+where
+    T: QueryFilter + InvertibleFilter,
+{
+    type Item<'a> = T::Item<'a>;
+
+    type Fetch<'a> = T::Fetch<'a>;
+
+    type State = T::State;
+
+    fn shrink<'wlong: 'wshort, 'wshort>(item: Self::Item<'wlong>) -> Self::Item<'wshort> {
+        T::shrink(item)
+    }
+
+    #[inline]
+    unsafe fn init_fetch<'w>(
+        world: UnsafeWorldCell<'w>,
+        state: &Self::State,
+        last_run: Tick,
+        this_run: Tick,
+    ) -> Self::Fetch<'w> {
+        T::init_fetch(world, state, last_run, this_run)
+    }
+
+    const IS_DENSE: bool = T::IS_DENSE;
+
+    #[inline]
+    unsafe fn set_archetype<'w>(
+        fetch: &mut Self::Fetch<'w>,
+        state: &Self::State,
+        archetype: &'w Archetype,
+        table: &'w Table,
+    ) {
+        T::set_archetype(fetch, state, archetype, table);
+    }
+
+    #[inline]
+    unsafe fn set_table<'w>(fetch: &mut Self::Fetch<'w>, state: &Self::State, table: &'w Table) {
+        T::set_table(fetch, state, table);
+    }
+
+    #[inline(always)]
+    unsafe fn fetch<'w>(
+        fetch: &mut Self::Fetch<'w>,
+        entity: Entity,
+        table_row: TableRow,
+    ) -> Self::Item<'w> {
+        T::fetch(fetch, entity, table_row)
+    }
+
+    fn update_component_access(state: &Self::State, access: &mut FilteredAccess<ComponentId>) {
+        T::update_component_access(state, access);
+
+        for filter in &mut access.filter_sets {
+            std::mem::swap(&mut filter.with, &mut filter.without);
+        }
```

Comment:
Pretty sure this is wrong:

- The current `Not` filter might be in a tuple (e.g. `(With<A>, Not<With<B>>)`), in which case `filter` will contain the access of the previous elements of the tuple (`With<A>` in the previous example) and those must not be flipped, which this code will instead do. This means that `(With<A>, Not<With<B>>)` will actually become equivalent to `(Without<A>, Without<B>)`, which is wrong, and even worse this won't be consistent with the archetypes actually fetched, which is unsound. For example this test fails:

```rs
    #[test]
    fn query_filter_not_bugged() {
        let mut world = World::default();
        let _e1 = world.spawn((A(0), B(0), C)).id();

        use crate::prelude::Query;
        fn system(
            q1: Query<(Entity, &mut C), (With<A>, Not<Without<B>>)>,
            q2: Query<(Entity, &mut C), With<A>>,
        ) {
            for (e1, _) in &q1 {
                for (e2, _) in &q2 {
                    // Surely two queries that both fetch `&mut C` can't yield the same entity, right?
                    assert_ne!(e1, e2);
                }
            }
        }

        let system_id = world.register_system(system);
        _ = world.run_system(system_id);
    }
```

- Swapping "positive" and "negative" terms doesn't work when negating a formula, it only works if the formula is made up of exactly one term (which is the case for `With` and `Without`, but not the case for more general filters like tuples and `Or`). The general algorithm for a DNF formula requires doing the cartesian product between the (negated) terms in each conjunction. For example `(A & !B & C) | (D & !E)` would become `(!A & !D) | (!A & E) | (B & !D) | (B & E) | (!C & !D) | (!C & E)` (yes this is exponential in the number of conjunctions)

A more robust approach would be to:
- build `T`'s component access in its own `FilteredAccess`
- negate that using the algorithm for general `DNF` formulas
- `and` the result with `filter`

The last two steps could be done in one step to improve the efficiency.